README.md
# chameleon-chromosomes
# Microchromosome fusions underpin convergent evolution of chameleon karyotypes

Author(s): [Natalie Cooper](mailto:natalie.cooper.@nhm.ac.uk)

THIS IS A WORK IN PROGRESS!

This repository contains all the code and some data used in the [paper](TO ADD link). 

To cite the paper: 
> Marcello Mezzasalma, Jeffrey W. Streicher, Fabio M. Guarino, Marc E.H. Jones, Simon P. Loader, Gaetano Odierna, and Natalie Cooper 2022. Microchromosome fusions underpin convergent evolution of chameleon karyotypes. TBC.

To cite this repo: 
> Natalie Cooper. 2023. GitHub: nhcooper123/chameleon-chromosomes: code for the paper. Zenodo. DOI: TO ADD.

![alt text](https://github.com/nhcooper123/chameleon-chromosomes/raw/master/manuscript/figures/new-tree-plot4.png)

## Data
All raw and cleaned data are available from the [NHM Data Portal](TO ADD).

For reproducibility purposes, if you want to rerun the data wrangling steps in script `analyses\01-data-wrangling`, download `CHROMO_STATES_2020-10-16.csv` and place it into a `raw-data/` folder along with the dataset of Meiri 2018: https://onlinelibrary.wiley.com/doi/full/10.1111/geb.12773.

The cleaned datasets, along rest of the ancillary data required to run all analyses and produce all figures is within the `data` folder.

* *Bayesian-tree-species.tre* - tree needed for the main analyses in the paper
* *chromosome-data-species.csv* - data needed for the main analyses in the paper
* *Bayesian-tree.tre* - tree for the all OTUs analyses in the supplementary materials
* *chromosome-data.csv* - data needed for the all OTUs analyses in the supplementary materials
* *chromevol-simulations-outputs-best.csv* - outputs from the "best" model of chromosome evolution used in the main text. The equivalent for the all OTUs analyses is in the folder `data/OTUs`.
* *null-simulations-outputs.csv* - outputs from the null simulations using the best model of chromosome evolution used in the main text. The equivalent for the all OTUs analyses is in the folder `data/OTUs`.

If you use the cleaned data please cite as follows: 
> Marcello Mezzasalma, Jeffrey W. Streicher, Fabio M. Guarino, Marc E.H. Jones, Simon P. Loader, Gaetano Odierna, and Natalie Cooper 2022. CHROMREP [Data set]. Natural History Museum. https://doi.org/10.5519/rmaopg7d.

The life-history data all comes from Meiri 2018 so please cite this paper if you use it.

> Meiri, S. Traits of lizards of the world: Variation around a successful evolutionary design. Global Ecol Biogeogr. 2018; 27: 1168– 1172

## ChromEvol analyses (`chromEvol/` folder)

The majority of analyses run in R, but we also used ChromEvol, a stand alone package available from http://chromevol.tau.ac.il/. The associated paper for the package is:

> Lior Glick, Itay Mayrose, ChromEvol: Assessing the Pattern of Chromosome Number Evolution and the Inference of Polyploidy along a Phylogeny, Molecular Biology and Evolution, Volume 31, Issue 7, July 2014, Pages 1914–1922, https://doi.org/10.1093/molbev/msu122   

To run these analyses you will need to download ChromEvol and save it in the `chromEvol/` folder. This unzips into a folder called `chromEvol_source/` that contains the package as an executable called `chromEvol`. I have not uploaded this to the repo because it's not my package. 

ChromEvol analyses require a dataset, a tree and a control file. These are within the `chromEvol/` folder preceded by `data_for_chromevol` or `tree_for_chroevol` or with the word `control` somewhere in their name respectively. Several versions exist for the different sensitivity analyses. These analyses save their outputs into the `chromEvol/results` folder, in folders named based on the names of the control files. Full details of how to build control files is in the ChromEvol manual at http://chromevol.tau.ac.il/. We provide code for creating appropriate datasets and trees in `analyses/02-`

Running chromEvol analyses is simple. Just open a Terminal in the `chromEvol` folder and run:

`chromEvol_source/chromEvol name-of-control-file`

You can run all analyses from our paper using the shell script: `run_all_models.sh`. 

The `chromEvol/` folder also includes a folder called `root_freq`. This has the three different root frequency definitions used in the different ChromEvol analyses (see text).

Finally the `OTUs` folder within `chromEvol/` contain the same things but for analyses using all OTUs (found in the supplementary materials of paper).

## ChromoSSE analyses (`ChromoSSE/` folder)

-------
## R analyses
The analysis code is divided into `.Rmd` files that run the analyses and plot the figures for each section of the paper/supplementary materials. 

Note that several of these scripts use an R package written by N.Cusimano. You can download this here https://www.en.sysbot.bio.lmu.de/people/employees/cusimano/use_r/. The easiest way to use this is to download the package and save the `ChromEvol.R` file. You can then install these functions in R using `source`. In my repo I kept this within the `chromEvol` folder so used the code `source("chromEvol/ChromEvol.R")` to install the functions.

1. **00-fix-tip-labels-bayesian-tree.R** changes the tip labels of the Bayesian tree to match the dataset.
1. **01-data-wrangling.Rmd** wrangles the data into a clean format for analyses.
2. **02A-data-for-chromosome-evolution-models_main.Rmd** creates data and tree files for the ChromEvol analyses.
3. **02B-data-for-chromosome-evolution-models_OTUs.Rmd** as above but for all OTUs analyses.
4. **03A-visualising-data-on-phylogenies_main.Rmd** visualises chromosome properties onto a phylogeny.
5. **03B-visualising-data-on-phylogenies_OTUs.Rmd** as above but for all OTUs analyses.
6. **04A-chromosome-evolution-models.Rmd** extract the outputs from the ChromEvol models, but only for the two main models of interest (constant rates and linear rates).
7. **04A-supp-chromosome-evolution-models-all_main.Rmd** extract the outputs from the ChromEvol models for all models.
8. **04B-chromosome-evolution-models_OTUs.Rmd** as 4A above but for all OTUs analyses.
9. **04B-supp-chromosome-evolution-models-all_OTUs.Rmd** as 4A-supp above but for all OTUs analyses.
10. **05A-micro-macro-correlates-analyses_main.Rmd** GLMs of haploid numbers of chromosomes against number of micro- and macro-chromosomes, numbers of micro- vs numbers of micro-chromosomes, ITS versus haploid numbers of chromosomes, and number of micro- and macro-chromosomes. Plus figures.
11. **05B-micro-macro-correlates-analyses_OTUs.Rmd** as above but for all OTUs analyses.
12. **06A-run-or-collate-simulations_main.Rmd** collates simulations outputs from ChromEvol and runs null simulations of haploid chromosome numbers.
13. **06B-run-or-collate-simulations_OTUs.Rmd** as above but for all OTUs analyses.
14. **06C-figures-simulation-models_main.Rmd** creates figures using simulations.
15. **06D-figures-simulation-models_OTUs.Rmd** as above but for all OTUs analyses.
16. **07A-taxon-species-pairs_main.Rmd** creates plots of species chromosome number distance versus phylogenetic distance.
17. **07B-taxon-species-pairs_OTUs.Rmd** as above but for all OTUs analyses.
18. **08A-ecological-correlates-figures.Rmd** creates figures of haploid chromosome number and a range of ecological, life-history and geographical variable for each species. FYI there are no all OTUs equivalents of these analyses as data are only available at the species level.
19. **08B-ecological-correlates-analyses.Rmd** MCMCglmm models of haploid chromosome number and a range of ecological, life-history and geographical variable for each species. FYI there are no all OTUs equivalents of these analyses as data are only available at the species level.
20. **09A-morphology-chromosome-no_main.Rmd** correlations of different aspects of chromosome morphology with haploid number of chromosomes.
21. **09B-morphology-chromosome-no_OTUs.Rmd** as above but for all OTUs analyses.
22. **F01-figure-tree-chromosomes.R** - creates figure 1 from ChromEvol outputs.
23. **figure-tree-chromosomes_main.R** - creates several versions of figure 1 from ChromEvol outputs.

-------
## Other folders

* `/outputs` contains the figures and tables
* `/manuscript` contains the manuscript materials in LaTeX format

-------
## Session Info
For reproducibility purposes, here is the output of `devtools::session_info()` used to perform the analyses in the publication.

## Checkpoint for reproducibility
To rerun all the code with packages as they existed on CRAN at time of our analyses we recommend using the `checkpoint` package, and running this code prior to the analysis:

```{r}
checkpoint("2022-09-10")
```



ChromoSSE/history.txt
# Each cladogenetic event type is assigned its own speciation rate.
# We set the rate of demi-polyploidization to 0.0 for simplicity.
clado_no_change ~ dnExponential(speciation_pr)
clado_fission ~ dnExponential(speciation_pr)
clado_fusion ~ dnExponential(speciation_pr)
clado_polyploid ~ dnExponential(speciation_pr)
clado_demipoly <- 0.0
# MCMC moves for speciation rates
moves[mvi++] = mvScale(clado_no_change, lambda=5.0, weight=1)
moves[mvi++] = mvScale(clado_fission, lambda=5.0, weight=1)
moves[mvi++] = mvScale(clado_fusion, lambda=5.0, weight=1)
moves[mvi++] = mvScale(clado_polyploid, lambda=5.0, weight=1)
# set a vector to hold the speciation rates
speciation_rates := [clado_no_change, clado_fission, clado_fusion, clado_polyploid, clado_demipoly]
total_speciation := sum(speciation_rates)
# map the speciation rates to chromosome cladogenetic events
clado_matrix := fnChromosomesCladoEventsBD(speciation_rates, max_chromo)
# a stochastic variable to represent the turnover. 
turnover ~ dnUniform(0, 1.0)
turnover.setValue(0.4)
moves[mvi++] = mvScale(turnover, lambda=5.0, weight=3.0)
# Make a vector of extinction rates for each state.
# In the basic ChromoSSE model we assume all chromosome 
# numbers have the same extinction rate.
for (i in 1:(max_chromo + 1)) {
    extinction[i] := turnover * total_speciation 
}
# Here we assume the frequency of chromosome numbers at the root of the tree
# are equal. In future examples we will instead treat the root frequencies
# as a free variable and estimate them from the observed data.
root_frequencies := simplex(rep(1, max_chromo + 1))
# Here we set the probability of sampling species at the present.
# We artificially use 1.0, but you should experiment with more realistic settings.
rho_bd <- 1.0
# Construct a variable for the tree and chromosomes drawn from a 
# cladogenetic character state dependent birth death process
chromo_bdp ~ dnCDBDP( rootAge            = phylogeny.rootAge(),
                      cladoEventMap      = clado_matrix,
                      extinctionRates    = extinction, 
                      Q                  = Q,
                      pi                 = root_frequencies,
                      rho                = rho_bd )
# and clamp the observed tree and the chromosome count data
chromo_bdp.clamp(phylogeny)
chromo_bdp.clampCharData(chromo_data)
# Finally we wrap our model into a single model object.
mymodel = model(phylogeny)
#########################
# MCMC
#########################
# Create the MCMC monitors:
# First, a screen monitor with some useful variables:
monitors[1] = mnScreen(printgen=10)
# The ancestral state monitor for sampling ancestral states:
monitors[2] = mnJointConditionalAncestralState(filename="output/ChromoSSE_anc_states.log", printgen=10, tree=phylogeny, cdbdp=chromo_bdp, withStartStates=true, type="NaturalNumbers")
# And another monitor for all the model variables:
monitors[3] = mnModel(filename="output/ChromoSSE_model.log", printgen=10)
# Now set up the MCMC.
mymcmc = mcmc(mymodel, monitors, moves)
# Run the MCMC. Note that for a real analysis you'd want to run many more 
# iterations and check for convergence.
mymcmc.run(200)
#########################
# Summarize the results 
#########################
# Now let's summarize the sampled ancestral chromosome numbers.
# First, read in the ancestral state trace:
anc_state_trace = readAncestralStateTrace("output/ChromoSSE_anc_states.log")
# Finally, summarize the values from the traces over the phylogeny.
# Here we do a marginal reconstruction of the ancestral states, discarding the first 25% of samples
# as burnin.
ancestralStateTree(phylogeny, anc_state_trace, "output/ChromoSSE_final.tree", burnin=0.25, include_start_states=true, reconstruction="marginal")
#########################################################################################################
#
# Open ChromoSSE_final.tree in FigTree and look at the ancestral state node labels.
# The 3 states with the highest posterior probabilities are given, along with their marginal
# posterior probabilities.
#
# Finally, run the plot_ChromoSSE.R script to generate a PDF plot of the ancestral states
# using the RevGadgets R package.
#
#########################################################################################################
phylogeny <- readTrees("data/aristolochia-bd.tree")[1]
phylogeny <- readTrees("data-example/aristolochia-bd.tree")[1]
phylogeny <- readTrees("./data-example/aristolochia-bd.tree")[1]
phylogeny <- readTrees("../data-example/aristolochia-bd.tree")[1]
phylogeny <- readTrees("~/Projects/chameleon-chromosomes/data-example/aristolochia-bd.tree")[1]
path
setwd("/Users/nataliecooper/Documents/Projects/chameleon-chromosomes")
setwd("Documents/Projects/chameleon-chromosomes")
setwd("Documents/Projects/chameleon-chromosomes/ChromoSSE")
source("ChromoSSE_linear_n18.Rev")
setwd("Documents/Projects/chameleon-chromosomes/ChromoSSE")
#######################
# Set working directory so RevBayes knows where to look for the data
setwd("Documents/Projects/chameleon-chromosomes/ChromoSSE")
# Decide number of iterations
iterations = 10000
### This is the best fitting model from ChromEvol###
source("ChromoSSE_exclude_n18.Rev")


analyses/figure-tree-chromosomes_main.R
# Make phylogeny figure(s)

# Load libraries
library(tidyverse)
library(here)
library(phytools)
source(here("chromEvol/ChromEvol.R")) #install this from source
library(ggtree)

# Read in the results for the best fitting models
results <- read.ce(here("chromEvol/results/02D_18-root-exclude_out/CONST_RATE_NO_DUPL"))

# Read in haploid number data
taxa <- read_csv(here("data/chromosome-data-species.csv"))
taxa <- filter(taxa, tips != "Riebrev" & tips != "Leiolep")

# Read in the tree
mytree <- read.tree(here("chromEvol/tree_for_chromevol_exclude"))

# Reorder data so it is the same order as the tree tip labels
taxa <- taxa[match(mytree$tip.label, taxa$tips), ]
# Replace tips with Binomials
mytree$tip.label <- taxa$Binomial
#---------------------------------------------------------
# Prep for plotting
#-----------------------------------------------------------
# Extract required data for plotting from ChromEvol object
# pp gives the number with the highest posterior probability
taxa_anc <- results[[1]][[1]]
taxa_tip <- results[[1]][[2]]
taxa_pp <- results[[1]][[3]]

# Create list of colours (colourblind friendly from: https://personal.sron.nl/~pault/#sec:qualitative)
mycolours <- c("#332288", "#33BBEE", "#44AA99", "#117733",
               "#62BB35", "#FDAE33", "#CC6677", "#EE3377", 
               "#AA4499", "#AA71FF")

taxa <- 
  taxa %>%
  mutate(colours = case_when(haploidn == 10 ~ "#332288",
                             haploidn == 11 ~ "#33BBEE",
                             haploidn == 12 ~ "#44AA99",
                             haploidn == 13 ~ "#117733",
                             haploidn == 14 ~ "#62BB35",
                             haploidn == 15 ~ "#FDAE33",
                             haploidn == 16 ~ "#CC6677",
                             haploidn == 17 ~ "#EE3377",
                             haploidn == 18 ~ "#AA4499",
                             haploidn == 19 ~ "#AA71FF"))

taxa_pp <- 
  data.frame(taxa_pp) %>%
  mutate(colours = case_when(taxa_pp == 10 ~ "#332288",
                             taxa_pp == 11 ~ "#33BBEE",
                             taxa_pp == 12 ~ "#44AA99",
                             taxa_pp == 13 ~ "#117733",
                             taxa_pp == 14 ~ "#62BB35",
                             taxa_pp == 15 ~ "#FDAE33",
                             taxa_pp == 16 ~ "#CC6677",
                             taxa_pp == 17 ~ "#EE3377",
                             taxa_pp == 18 ~ "#AA4499",
                             taxa_pp == 19 ~ "#AA71FF"))
#-----------------------------------------------------------
# plot (minus pies)
plot(mytree, 
     show.tip.label = TRUE, no.margin = TRUE, 
     tip.color = taxa$colours, cex = 0.5, font = 3, 
     label.offset = 5)
tiplabels(pch = 16, col = taxa$colours, offset = 0.5)
tiplabels(text = taxa$haploidn, frame = "none", cex = 0.5, 
         offset = 3)
#nodelabels(pie = taxa_anc, piecol = mycolours, cex = 0.6)
nodelabels(pch = 16, col = taxa_pp$colours, cex = 3)
nodelabels(text = taxa_pp$taxa_pp, frame = "circle", bg = "white", cex = 0.5)

# Save the plot using the export function once it is the correct size.
# width 600/700
#-----------------------------------------------------------
# No colours for tips
plot(mytree, 
     show.tip.label = TRUE, no.margin = TRUE, 
     tip.color = "black", cex = 0.5, font = 3, 
     label.offset = 5)
tiplabels(pch = 16, col = taxa$colours, offset = 0.5)
tiplabels(text = taxa$haploidn, frame = "none", cex = 0.5, 
          offset = 3)
#nodelabels(pie = taxa_anc, piecol = mycolours, cex = 0.6)
nodelabels(pch = 16, col = taxa_pp$colours, cex = 3)
nodelabels(text = taxa_pp$taxa_pp, frame = "circle", bg = "white", cex = 0.5)
#-----------------------------------------------------------
# FAN
par(mar = c(0.01, .01, .01, .01))
plot(mytree, 
     show.tip.label = TRUE, no.margin = FALSE, 
     tip.color = taxa$colours, cex = 0.7, font = 3, 
     label.offset = 5, type = "fan")
tiplabels(pch = 16, col = taxa$colours, offset = 0.5)
tiplabels(text = taxa$haploidn, frame = "none", cex = 0.5, 
          offset = 3)
#nodelabels(pie = taxa_anc, piecol = mycolours, cex = 0.6)
nodelabels(pch = 16, col = taxa_pp$colours, cex = 4)
nodelabels(text = taxa_pp$taxa_pp, frame = "circle", bg = "white", cex = 0.7)
# Width = 950
#-----------------------------------------------------------
# With pies

# Numbers 1-9 grey, 10-19 colours, 20-29 - grey.
mycolours_pies <- c(rep("#DDDDDD",9),
                "#332288", "#33BBEE", "#44AA99", "#117733",
               "#62BB35", "#FDAE33", "#CC6677", "#EE3377", 
               "#AA4499", "#AA71FF", rep("#DDDDDD", 10))

# Fan
plot(mytree, type = "fan",
     show.tip.label = TRUE, no.margin = FALSE, cex = 0.7, font = 3, 
     label.offset = 10, tip.color = taxa$colours)
tiplabels(pie = taxa_tip, piecol = mycolours_pies, offset = 5, cex = 0.6)
tiplabels(text = taxa$haploidn, frame = "none", cex = 0.6, 
          offset = 5)
nodelabels(pie = taxa_anc, piecol = mycolours_pies, cex = 0.8)
nodelabels(text = taxa_pp$taxa_pp, frame = "circle", bg = "white", cex = 0.6)

# Save the plot using the export function once it is the correct size.
# THIS IS MY FAVOURITE PLOT
plot(mytree, type = "fan",
     show.tip.label = TRUE, no.margin = FALSE, cex = 0.7, font = 3, 
     label.offset = 10)
tiplabels(pie = taxa_tip, piecol = mycolours_pies, offset = 5, cex = 0.6)
tiplabels(text = taxa$haploidn, frame = "none", cex = 0.6, 
          offset = 5)
nodelabels(pie = taxa_anc, piecol = mycolours_pies, cex = 0.8)
nodelabels(text = taxa_pp$taxa_pp, frame = "circle", bg = "white", cex = 0.6)